### PR TITLE
Improvements on "Each" rule and test classes

### PR DESCRIPTION
**Pull request by batusa**
regarding issue #598

### CODE REVIEW

**Comment by batusa** [comment]
now `Each` rule has 100% of the code covered :)

@henriquemoody anything else I can improve in this PR? feel free to say!

**Comment by henriquemoody** [comment]
The only thing I can point out is that you're using a lot of other rules instead of mocks, that's one of problems described on #379. 

You can use the [WhenTest](https://github.com/Respect/Validation/blob/master/tests/unit/Rules/WhenTest.php) as a reference or I can just merge it as is, it's your call.

File: tests/unit/Rules/EachTest.php [hunk 1]
```diff
@@ -16,13 +16,52 @@
  * @covers Respect\Validation\Rules\Each
  * @covers Respect\Validation\Exceptions\EachException
  */
-class EachTest extends \PHPUnit_Framework_TestCase
+class EachTest extends RuleTestCase
 {
+    public function providerForValidInput()
+    {
+        $ruleNotEmpty = new Each(new NotEmpty());
```
Comment (batusa) [inline tests/unit/Rules/EachTest.php]: here I'd do something like:

```php
$ruleNotEmpty = new Each($this->getRuleMock(true, 'NotEmpty'));
```

?

File: tests/unit/Rules/EachTest.php [hunk 2]
```diff
@@ -16,13 +16,52 @@
...
+            [$ruleNotEmpty, $intStack],
+            [$ruleNotEmpty, $stdClass],
+            [$ruleAlphaItemIntKey, ['a', 'b', 'c', 'd', 'e']],
+            [$ruleOnlyKeyValidation, ['a', 'b', 'c', 'd', 'e']],
+        ];
+    }
+
+    public function providerForInvalidInput()
+    {
+        $rule = new Each(new NotEmpty());
```
Comment (batusa) [inline tests/unit/Rules/EachTest.php]: and here would be like:

```php
$ruleNotEmpty = new Each($this->getRuleMock(false, 'NotEmpty'));
```

?

@henriquemoody

**Comment by batusa** [comment]
I created the mocks, check it out :)

File: tests/unit/Rules/EachTest.php [hunk 3]
```diff
@@ -16,13 +16,52 @@
...
  * @covers Respect\Validation\Exceptions\EachException
  */
-class EachTest extends \PHPUnit_Framework_TestCase
+class EachTest extends RuleTestCase
 {
+    public function providerForValidInput()
+    {
+        $ruleNotEmpty = new Each($this->getRuleMock(true, 'NotEmpty'));
+        $ruleAlphaItemIntKey = new Each($this->getRuleMock(true, 'Alpha'), $this->getRuleMock(true, 'IntVal'));
+        $ruleOnlyKeyValidation = new Each(null, $this->getRuleMock(true, 'IntVal'));
```
Comment (henriquemoody) [inline tests/unit/Rules/EachTest.php]: The names of the rules here is just irrelevante, it's just the name of the mock classes

Comment (batusa) [inline tests/unit/Rules/EachTest.php]: do I have to remove them?

Comment (henriquemoody) [inline tests/unit/Rules/EachTest.php]: It's okey

**Comment by henriquemoody** [comment]
Merged manually by 58461766e42369fecc16032d0e322970c3512b0e

Thanks for contributing! :panda_face:

**Pull request closed without merging**